src/mlpack/core/util/sfinae_utility.hpp

```diff
@@ -15,6 +15,110 @@
 #ifndef MLPACK_CORE_SFINAE_UTILITY
 #define MLPACK_CORE_SFINAE_UTILITY
 
+#include <type_traits>
+
+namespace mlpack {
+namespace sfinae {
+
+/*
+ * MethodFormDetecter is a tool that helps to find out whether a given class has
+ * a method of the requested form. For that purpose MethodFormDetecter defines
+ * an operator() that accepts a class member pointer for the given class. If the
+ * operator()(&Class::Method) call can be compiled, then the given class has a
+ * method of the requested form. For any provided AdditionalArgsCount, the check
+ * succeeds only if the given class has exactly one method of the requested form
+ * with AdditionalArgsCount additional arguments.
+ *
+ * The tool is dedicated to be used in type functions (structs) generated by the
+ * macro HAS_METHOD_FORM.
+ *
+ * @tparam MethodForm A template class member pointer type to a method of the
+ *   form to look for.
+ * @tparam Class A class in which a method of the requested form should be
+ *   looked for.
+ * @tparam AdditionalArgsCount A number of additional arguments.
+ */
+template<typename Class,
+         template<typename...> class MethodForm,
+         size_t AdditionalArgsCount>
+struct MethodFormDetecter;
```

Comment:
Can you change the name to `MethodFormDetector` please?  Just a slight misspelling :)